### PR TITLE
BUG: preserve join keys dtype

### LINKED ISSUE #8596: BUG: DataFrame outer merge changes key columns from int64 to float64
```
import pandas as pd

df1 = pd.DataFrame({'key': [1,2,3,4], 'val1': [1,2,3,4]})
df2 = pd.DataFrame({'key': [1,2,3,5], 'val2': [1,2,3,4]})

df = df1.merge(df2, how='outer')
```

Was expecting `key` to stay int64, since a merge can't introduce missing key values if none were present in the inputs.

```
print df.dtypes

key     float64
val1    float64
val2    float64
dtype: object
```

Version 0.15.0-6-g403f38d

### PR DESCRIPTION
**Pull request by jreback**
- closes #8596, preserve join keys dtype
- adds `Index.where` method for all Index types (like `np.where/Series.where`), but preserves dtypes

### CODE REVIEW

File: pandas/tseries/period.py [hunk 1]
```diff
@@ -946,6 +946,27 @@ def repeat(self, n, *args, **kwargs):
         # overwrites method from DatetimeIndexOpsMixin
         return self._shallow_copy(self.values.repeat(n))
 
+    def where(self, cond, other=None):
+        """
+        .. versionadded:: 0.18.2
+
+        Return an Index of same shape as self and whose corresponding
+        entries are from self where cond is True and otherwise are from
+        other.
```
Comment (jreback) [inline pandas/tseries/period.py]: @sinhrks I wish that could get some more uniformity w.r.t. Periods. I realize that there is a bug here with ``other`` which I fixed in ``tseries/base.py/.where``, end up duplicating code here.

Comment (sinhrks) [inline pandas/tseries/period.py]: Does `_shallow_copy_with_infer` work?

- https://github.com/pydata/pandas/blob/master/pandas/indexes/base.py#L354
- https://github.com/pydata/pandas/blob/master/pandas/tseries/period.py#L288

Comment (jreback) [inline pandas/tseries/period.py]: its more complicated. I needed prelim logic different fro dti & period, so would have to have a helper function. We do this a fair amount, e.g. process for timezones, run a function, then reverse. Its ok the way we do it now, but it does duplicate *some* code.

**Comment by jreback** [comment]
ok, removed the dep on my other PR.

@sinhrks @jorisvandenbossche @TomAugspurger comments?

**Comment by jreback** [comment]
@sinhrks any comments on this?

**Pull request closed without merging**